### DISCUSSION TITLE
Custom layer disappears at zoom level 12 in globe projection

### DISCUSSION THREAD

**Discussion by kamil-sienkiewicz-asi** [maplibre/maplibre-gl-js#5237] [Q&A]
**maplibre-gl-js version**: 5.0.0-pre.10

**browser**: 131.0.6778.140 (Official Build) (arm64)

### Steps to Trigger Behavior

 1. Switch to globe projection
 2. Add custom layer
 3. Zoom to/past zoom level 12.0

https://codepen.io/kamil-sienkiewicz-asi/pen/bNbgPqR?editors=1010

### Expected Behavior

Custom layer does not disappear

### Actual Behavior

https://github.com/user-attachments/assets/f0ec1f3b-176d-4194-9fcb-f6ef07df3c4d

**Comment by HarelM** [comment]
Is the "shake" of the triangles expected as part of this demo? I find it odd...

**Comment by HarelM** [comment]
CC: @ibesora @kubapelc

**Comment by kubapelc** [comment]
The map switches to mercator projection around zoom 12. Looks like you are drawing vertices with coordinates computed with `toGlobeCoords` even after the switch to mercator, which uses a different coordinate system. See the `globe-custom-simple` (and to some extent `globe-3d-model.html` and `globe-custom-tiles.html`) for examples of how to draw custom layers that work for either projection.

If you want your layers to work well at high zooms, you will need to take floating point precision into account and for example draw things relative to a tile, using the tile's matrix (in mercator, globe has one matrix for everything). See `globe-custom-tiles.html`.

Also make sure to use the `2d` custom layer mode instead of `3d` when drawing flat layers. Using `3d` will make sure the depth buffer is filled, so that custom layers can be occluded by the rest of the map. The (single) triangle in your custom layer appears as 3 triangles, because it is clipping through the planet.

**Comment by kamil-sienkiewicz-asi** [comment]
@kubapelc isn't this switch to mercator at 12 zoom level a little bit misleading? In our use case we are setting an uniform of projection (using map.getProjection()),  based on that we are calculating ECEF or mercator coordinates in vertex shader. However when I zoom past 12.0 zoom level, `map.getProjection()` this returns `{type: 'globe'}` which results in disappearing layers. I could simply add check for map zoom and swap uniform, but that's kinda "magic" number that can change without my knowledge at next release.
